docs/reference/api-concepts.md

```diff
@@ -0,0 +1,232 @@
+        ---
+        200 OK
+        Content-Type: application/json
+        {
+          "kind": "PodList",
+          "apiVersion": "v1",
+          "metadata": {
+            "resourceVersion":"10245",
+            "continue": "", // continue token is empty because we have reached the end of the list
+            ...
+          },
+          "items": [...] // returns pods 1001-1253
+        }
+
+Note that the `resourceVersion` of the list remains constant across each request, indicating the server is showing us a consistent snapshot of the pods. Pods that are created, updated, or deleted after version `10245` would not be shown unless the user makes a list request without the `continue` token.  This allows clients to break large requests into smaller chunks and t hen perform a watch operation on the full set without missing any updates.
+
+
+## Alternate representations of resources
+
+By default Kubernetes returns objects serialized to JSON with content type `application/json`. This is the default serialization format for the API. However, clients may request the more efficient Protobuf representation of these objects for better performance at scale. The Kubernetes API implements standard HTTP content type negotation - passing an `Accept` header with a `GET` call will request that the server return objects in the provided content type, while sending an object in Protobuf to the server for a `PUT` or `POST` call takes the `Content-Type` header. The server will return a `Content-Type` header if the requested format is supported, or the `406 Not acceptable` error if an invalid content type is provided.
+
+See the API documentation for a list of supported content types for each API.
+
+For example:
+
+1. List all of the pods on a cluster in Protobuf format
+
+        GET /api/v1/pods
+        Accept: application/vnd.kubernetes.protobuf
+        ---
+        200 OK
+        Content-Type: application/vnd.kubernetes.protobuf
+        ... binary encoded PodList object
+
+2. Create a pod by sending Protobuf encoded data to the server, but request a response in JSON
+
+        POST /api/v1/namespaces/test/pods
+        Content-Type: application/vnd.kubernetes.protobuf
+        Accept: application/json
+        ... binary encoded Pod object
+        ---
+        200 OK
+        Content-Type: application/json
+        {
+          "kind": "Pod",
+          "apiVersion": "v1",
+          ...
+        }
+
+Not all API resource types will support Protobuf, specifically those defined via Custom Resource Definitions or those that are API extensions. Clients that must work against all resource types should specify multiple content types in their `Accept` header to support fallback to JSON:
+
+```
+Accept: application/vnd.kubernetes.protobuf, application/json
+```
+
+
+### Protobuf encoding
+
+Kubernetes uses an envelope wrapper to encode Protobuf responses, since unlike JSON a Protobuf message is not self-describing (the client cannot determine the type).
```

Comment:
The point you're trying to make here didn't immediately click with me.

Out of interest, why didn't we just require that TypeMeta always be field 1?

Reply:
It is by convention today, but we wanted to make it easy for a client to sniff content type as well for things on disk or in etcd.  Proto doesn't require field 1 to be at the beginning of a serialized object (at least, pretty sure it doesn't) so we wouldn't have been able to sniff for it.

Reply:
Oh, we also wanted to support gzipping in the future.